### PR TITLE
add debug log for patching/extracting i18n nodes

### PR DESCRIPTION
**Pull request by shimizukawa**
### Feature or Bugfix
- Feature for debugging

### Purpose
- ease to debug of i18n

### Detail
- To know what happen on `make gettext` and `make html` with using i18n feature.

for example:
```
sphinx-build -M gettext  . _build -vvv
...
[i18n] PATCH: 'docutils.nodes.field_name' to have source and line: <field_name>pep</field_name>
[app] emitting event: 'doctree-read'(<document: <docinfo...><target...><section "changes in sphinx; chan ...>,)
[i18n] SKIP 'sphinx.addnodes.versionmodified' because node is in IGNORED_NODES and no node['translatable']: <versionmodified type="versionadded" version="1.1"><paragraph translatable="Fals...
[i18n] SKIP 'docutils.nodes.paragraph' because not node['translatable']: <paragraph translatable="False"><inline classes="versionmodified">New in version...
[i18n] SKIP 'docutils.nodes.literal_block' because no node.source: <literal_block xml:space="preserve">.. autoclass:: my.Class
   :members:
   :pri...
```

### CODE REVIEW

File: sphinx/util/logging.py [hunk 1]
```diff
@@ -94,6 +94,18 @@ def convert_serializable(records):
             r.location = get_node_location(location)  # type: ignore
 
 
+def get_full_module_name(node):
+    return '{}.{}'.format(node.__module__, node.__class__.__name__)
+
+
+def get_domxml(node, length=80):
+    # text = node.asdom().toxml()  # crush if node has secnumber with tuple value
```
Comment (shimizukawa) [inline sphinx/util/logging.py]: I'll make another issue for this problem.

Comment (shimizukawa) [reply sphinx/util/logging.py]: Please refere https://github.com/sphinx-doc/sphinx/issues/4919

File: sphinx/util/logging.py [hunk 2]
```diff
@@ -94,6 +94,18 @@ def convert_serializable(records):
             r.location = get_node_location(location)  # type: ignore
 
 
+def get_full_module_name(node):
+    return '{}.{}'.format(node.__module__, node.__class__.__name__)
```
Comment (tk0miya) [inline sphinx/util/logging.py]: Please add typehints here.

File: sphinx/util/logging.py [hunk 3]
```diff
@@ -94,6 +94,18 @@ def convert_serializable(records):
...
+
+
+def get_domxml(node, length=80):
```
Comment (tk0miya) [inline sphinx/util/logging.py]: Here is also.

File: sphinx/util/nodes.py [hunk 4]
```diff
@@ -96,17 +104,30 @@ def is_translatable(node):
     if isinstance(node, addnodes.translatable):
         return True
 
+    if isinstance(node, nodes.Inline) and 'translatable' not in node:
+        # inline node must not be translated if 'translatable' is not set
+        return False
```
Comment (tk0miya) [inline sphinx/util/nodes.py:108]: This might not be related with this PR. Is this right?

**Review by tk0miya — CHANGES_REQUESTED** [review]

File: sphinx/util/nodes.py (hunk 4 above)
Comment (shimizukawa) [reply sphinx/util/nodes.py:108]: It relates this PR, sorry for my less comment on this change.

Previously, on line 116, the code returned False under the same conditions. Now ton of log outputs are emitted on line 117 for each Inline nodes, so I divided this condition into separated "if" statement.

File: sphinx/util/logging.py [hunk 5]
```diff
@@ -94,6 +94,36 @@ def convert_serializable(records):
...
+    '<paragraph translatable="False"><inline classes="versionmodified">New in version...'
+
+    :param nodes.Node node: target node
+    :param int length:
+       length of return value to be striped. if false-value is specified, repr_domxml
+       returns full of DOM XML representation.
+    :return: DOM XML representation
+    """
+    # text = node.asdom().toxml()  # #4919 crush if node has secnumber with tuple value
+    text = str(node) # workaround for #4919
```
Comment (shimizukawa) [inline sphinx/util/logging.py]: Please refere #4919

File: sphinx/util/nodes.py [hunk 6]
```diff
@@ -21,7 +21,7 @@
 
 if False:
     # For type annotation
-    from typing import Any, Callable, Iterable, List, Set, Tuple, Union  # NOQA
+    from typing import Any, Callable, Iterable, List, Set, Tuple  # NOQA
```
Comment (shimizukawa) [inline sphinx/util/nodes.py:24]: Union is not used.

**Comment by shimizukawa** [comment]
updated.

File: sphinx/util/logging.py [hunk 7]
```diff
@@ -94,6 +94,36 @@ def convert_serializable(records):
...
+    return DOM XML representation of the specified node like:
+    '<paragraph translatable="False"><inline classes="versionmodified">New in version...'
+
+    :param nodes.Node node: target node
+    :param int length:
+       length of return value to be striped. if false-value is specified, repr_domxml
+       returns full of DOM XML representation.
+    :return: DOM XML representation
+    """
+    # text = node.asdom().toxml()  # #4919 crush if node has secnumber with tuple value
```
Comment (tk0miya) [inline sphinx/util/logging.py:120]: Is this comment needed?

File: sphinx/util/nodes.py (hunk 4 above)
Comment (tk0miya) [reply sphinx/util/nodes.py:108]: Okay, I understand.

File: sphinx/util/logging.py [hunk 8]
```diff
@@ -94,6 +94,36 @@ def convert_serializable(records):
...
+    return full module dotted path like: 'docutils.nodes.paragraph'
+
+    :param nodes.Node node: target node
+    :return: full module dotted path
+    """
+    return '{}.{}'.format(node.__module__, node.__class__.__name__)
+
+
+def repr_domxml(node, length=80):
+    # type: (nodes.Node, Optional[int]) -> str
```
Comment (tk0miya) [inline sphinx/util/logging.py]: Is this really `str`? not `unicode`?

**Review by tk0miya — APPROVED** [review]
LGTM with nits

File: sphinx/util/logging.py (hunk 8 above)
Comment (shimizukawa) [reply sphinx/util/logging.py]: Ah, .. unicode is correct. I'll fix it.

File: sphinx/util/logging.py (hunk 7 above)
Comment (shimizukawa) [reply sphinx/util/logging.py:120]: After fixing #4919, this line must be commented-in and next 121 line will be removed.

Comment (tk0miya) [reply sphinx/util/logging.py:120]: I see

**Review by tk0miya — APPROVED** [review]
LGTM!

**Comment by shimizukawa** [comment]
Thanks!

File: sphinx/util/nodes.py [hunk 9]
```diff
@@ -41,13 +41,19 @@ def apply_source_workaround(node):
...
     if isinstance(node, nodes.classifier) and not node.rawsource:
+        logger.debug('[i18n] PATCH: %r to have source, line and rawsource: %s',
+                     logging.get_full_module_name(node), logging.repr_domxml(node))
         definition_list_item = node.parent
         node.source = definition_list_item.source
         node.line = definition_list_item.line - 1
         node.rawsource = node.astext()  # set 'classifier1' (or 'classifier2')
     if isinstance(node, nodes.image) and node.source is None:
+        logger.debug('[i18n] PATCH: %r to have source, line: %s',
+                     logging.get_full_module_name(node), logging.repr_domxml(node))
```
Comment (tk0miya) [inline sphinx/util/nodes.py:52]: @shimizukawa 
I don't know why I didn't think to ask this sooner, why `get_full_module_name()` and `repr_domxml()` are placed on  logging module? I think both are utilities for nodes. How about moving them to `sphinx.util.nodes`?

Comment (shimizukawa) [reply sphinx/util/nodes.py:52]: OK, I'll move it soon.

Comment (shimizukawa) [reply sphinx/util/nodes.py:52]: done. https://github.com/sphinx-doc/sphinx/commit/5fbc6d378dd9f95948501a564194f258bfb60174

**Pull request merged**